### PR TITLE
Add support for Resnext101 WSL features

### PR DESCRIPTION
**Pull request by klshuster**
**Patch description**
Add support for extracting/loading [Resnext WSL features](https://pytorch.org/hub/facebookresearch_WSL-Images_resnext/) (from PyTorch)

Access via specifying one of the following `--image-mode`s:
- resnext101_32x8d_wsl
- resnext101_32x16d_wsl
- resnext101_32x32d_wsl
- resnext101_32x48d_wsl

**Testing steps**
1. install latest pytorch and torchvision distributions
2. `python examples/display_data -t flickr30k --image-mode resnext101_32x8d_wsl -dt valid`

**Logs**
```
$ python examples/extract_image_feature.py -t flickr30k -dt valid --image-mode resnext101_32x8d_wsl
...
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1014/1014 [00:01<00:00, 560.29ex/s]
[ Finished extracting images ]
```

### CODE REVIEW

File: parlai/core/image_featurizers.py [hunk 1]
```diff
@@ -109,7 +131,8 @@ def extract(self, image, path=None):
         transform = self.transform(image).unsqueeze(0)
         if self.use_cuda:
             transform = transform.cuda()
-        feature = self.netCNN(transform)
+        with self.torch.no_grad():
+            feature = self.netCNN(transform)
```
Comment (stephenroller) [inline parlai/core/image_featurizers.py]: why the nograd? MAYBE I WANNA FINETUNE MY 900M PARAMETER MODEL, HUH KURT?

File: parlai/scripts/extract_image_feature.py [hunk 2]
```diff
@@ -103,10 +103,13 @@ def extract_feats(opt):
         world = create_task(opt, agent)
 
         total_exs = world.num_examples()
-        # TODO: wrap in a tqdm
+        pbar = tqdm.tqdm(unit='ex', total=total_exs)
         while not world.epoch_done():
             world.parley()
+            pbar.update()
```
Comment (stephenroller) [inline parlai/scripts/extract_image_feature.py]: this idiom appears a bunch of places. is there a better improvement? no action required here but

File: parlai/core/image_featurizers.py [hunk 3]
```diff
@@ -57,13 +57,35 @@ def init_cnn(self, opt):
...
+            CNN = getattr(torchvision.models, cnn_type)
+
+            # cut off the additional layer.
+            self.netCNN = nn.Sequential(
+                *list(CNN(pretrained=True).children())[:layer_num])
+        elif 'resnext' in self.image_mode:
+            try:
+                model = torch.hub.load('facebookresearch/WSL-Images', self.image_mode)
+                self.netCNN = nn.Sequential(
+                    *list(model.children())[:-1])
```
Comment (stephenroller) [inline parlai/core/image_featurizers.py]: leave a comment saying you're cutting off the classification layer?

File: parlai/core/image_featurizers.py [hunk 4]
```diff
@@ -57,13 +57,35 @@ def init_cnn(self, opt):
...
+                # Perhaps specified one of the wrong model names
+                print('If you have specified one of the resnext101 wsl models, '
+                      'please make sure it is one of the following: \n'
+                      'resnext101_32x8d_wsl, resnext101_32x16d_wsl, '
+                      'resnext101_32x32d_wsl, resnext101_32x48d_wsl')
+                raise e
+            except AttributeError:
+                # E.g. "module 'torch' has no attribute 'hub'"
+                raise RuntimeError(
+                    'Please install the latest pytorch distribution to have access '
```
Comment (stephenroller) [inline parlai/core/image_featurizers.py]: Do you know what version needs it? It'd be nice to say it explicitly.

File: parlai/core/image_featurizers.py [hunk 5]
```diff
@@ -57,13 +57,35 @@ def init_cnn(self, opt):
...
+                    'to the resnext101 wsl models')
+            except Exception as e:
```
Comment (stephenroller) [inline parlai/core/image_featurizers.py]: you can delete these two lines; no need to catch and raise.

File: parlai/core/image_featurizers.py [hunk 6]
```diff
@@ -57,13 +57,35 @@ def init_cnn(self, opt):
...
             torch.cuda.set_device(opt.get('gpu', -1))
 
-        cnn_type, layer_num = self._image_mode_switcher()
-
-        # initialize the pretrained CNN using pytorch.
-        CNN = getattr(torchvision.models, cnn_type)
-
-        # cut off the additional layer.
-        self.netCNN = nn.Sequential(*list(CNN(pretrained=True).children())[:layer_num])
+        if 'resnet' in self.image_mode:
```
Comment (stephenroller) [inline parlai/core/image_featurizers.py]: resnet vs resnext, so subtle lol.

does it make sense to refactor each of these load processes into some helper methods?

File: parlai/core/image_featurizers.py (hunk 1 above)
Comment (stephenroller) [reply parlai/core/image_featurizers.py]: (real question though.)

Comment (klshuster) [reply parlai/core/image_featurizers.py]: Speeds up computation; also recommended [here](https://pytorch.org/hub/facebookresearch_WSL-Images_resnext/) (see third code block)

File: parlai/core/image_featurizers.py (hunk 3 above)
Comment (klshuster) [reply parlai/core/image_featurizers.py]: good idea

File: parlai/core/image_featurizers.py (hunk 6 above)
Comment (klshuster) [reply parlai/core/image_featurizers.py]: yes, I will do so

File: parlai/core/image_featurizers.py (hunk 4 above)
Comment (klshuster) [reply parlai/core/image_featurizers.py]: yes

File: parlai/core/image_featurizers.py [hunk 7]
```diff
@@ -25,57 +25,98 @@ def __init__(self, opt):
...
-
-        # initialize the pretrained CNN using pytorch.
-        CNN = getattr(torchvision.models, cnn_type)
-
-        # cut off the additional layer.
-        self.netCNN = nn.Sequential(*list(CNN(pretrained=True).children())[:layer_num])
+            torch.cuda.set_device(self.opt.get('gpu', -1))
+        self.torch = torch
+        self.torchvision = torchvision
+        self.transforms = transforms
```
Comment (samhumeau) [inline parlai/core/image_featurizers.py]: interesting pattern

**Comment by samhumeau** [comment]
It's awesome that those are finally open source. They are so much better than the vanilla resnet 152

File: parlai/core/image_featurizers.py (hunk 7 above)
Comment (klshuster) [reply parlai/core/image_featurizers.py]: this script is written as platform agnostic - i.e., it does not require that you have torch installed. So we do not import torch until someone specifies actual features to extract

**Comment by chengxili** [comment]
Just want to comment about seems " python examples/extract_image_feature.py" extracted me all the features while display_data seems not.

**Pull request merged**